### DISCUSSION TITLE
Geometric Calibration

### DISCUSSION THREAD

**Discussion by kylechampley** [llnl/LEAP#54] [General]
Geometric calibration has been an active topic on the LEAP github page for the past several weeks.  I am starting this discussion for this topic.  If you know of any methods or tools for geometric calibration, please list them here.  If they are general-purpose methods, I would be encouraged to add them to LEAP.  So far, I have implemented most of what I am familiar with.

A few years ago, I can across the following paper.  It has been a while since I read it, but I remember that when I read it, I thought it was a good method for geometric calibration.
[Joint iterative reconstruction and 3D rigid alignment for X-ray tomography](https://opg.optica.org/oe/fulltext.cfm?uri=oe-30-6-8898).

Let me know what you think.  The following people may be interested in this discussion
@lc82111
@ProjX0
@hws203
@andrewerf

**Comment by jeffhwang02** [comment]
Below paper and their open code have the FDCTcalibraion which may be useful for  Leap new feature.


https://acrobat.adobe.com/id/urn:aaid:sc:AP:f01d4516-e397-42bf-ab87-e293ccff1716

**Comment by hws203** [comment]
(https://ieeexplore.ieee.org/document/9364757)

Above paper needs to be checked too.

**Comment by lc82111** [comment]
I would like to apply geometric corrections to each individual projection instead of using the same correction for all projections.

**Reply by hws203** [reply]
If such individual correction works well, then that could be very useful. I have several kinds of sample images for such testing. If it is possible, please let me know it.

**Comment by hws203** [comment]
Based on my experience of large size detector, each row's  sinogram based calibration may be necessary too.

**Comment by hws203** [comment]
https://www.mdpi.com/2431348

![cal](https://github.com/LLNL/LEAP/assets/73321363/e02fa16e-0af5-4b32-9eaa-98fef3472336)

**Comment by kylechampley** [comment]
@jeffhwang02 that reference you gave seem like a really good algorithm.  I check the [link](https://www5.cs.fau.de/research/software/epipolar-consistency/) in the paper and it leads to a web page with some materials and a link to the source code.  The paper is very hard to follow and the source code is quite long, so I can't make sense of it.  My hope is that this code works well and can be used as a preprocessing step.  I'd like this kind of functionality in LEAP, but at this moment I don't have time to incorporate this.

**Comment by kylechampley** [comment]
I think the paper @hws203 post above (Distributed Optimization for Nonrigid Nano-Tomography) may be able to be used for calibration of each projection, but I see two possible issues.  First of all, that paper was written for parallel-beam CT, so extending it to cone-beam CT would require some extra work.  And second of all, the joint reconstruction + geometric calibration algorithms don't always work.  The job of iterative reconstruction is to fit the model (i.e., the reconstruction) to the data.  So even when the geometry is wrong, iterative reconstruction will attempt to find a reconstruction that agrees with the misaligned data.

**Reply by hws203** [reply]
Thanks for your comments, How about upgrading your consistency-cost function to work even with some strong noise level?   I did check your consistency-cost with the large level of noise at your original demo script. And I could see that it shows very fluctuated returns whenever I did same script.  It looks noise sensitive.

**Reply by kylechampley** [reply]
The consistency cost is based on the zeroth order consistency condition which is a sum of an entire column of data and thus the method should be very robust to noise.

If you are noticing that you are getting inaccurate results due to high noise, then just run the data through one of LEAP's denoising algorithms on the projection data prior to running the calibration routine.

**Reply by jeffhwang02** [reply]
If you use one line of colunm then this entire line data may be weak point. As you can see at the sample of my cylinder sample case. There are almost empty data at the center line of colunm. So I have to use the lower line colunm for cost function. Then it shows stable result. It is related to the sample shape. An entire column of data + partial row of data summing may be more robust.

**Comment by hws203** [comment]
![AI-train_Prj](https://github.com/LLNL/LEAP/assets/73321363/7f76e40e-99dd-4aaa-a9e5-da101bafa3d5)


Above block diagram approaching looks good too.

from : Perspective Projection-Based 3D CT Reconstruction from Biplanar X-rays (ICASSP 2023, Best Student Paper Award).

**Comment by hws203** [comment]
@kylechampley  I have checked your new estimate_tilt() function of v1.18, and I can see that function works well with my own real data.
And that function is fast too. That could be useful for real data fine reconstruction.
